libraries/AP_UAVCAN/AP_UAVCAN_Servers.cpp

```diff
@@ -0,0 +1,304 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Siddharth Bharat Purohit
+ */
+
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "AP_UAVCAN_Servers.h"
+
+#ifdef HAS_UAVCAN_SERVERS
+
+#include <uavcan/protocol/dynamic_node_id_server/event.hpp>
+#include <uavcan/protocol/dynamic_node_id_server/storage_backend.hpp>
+#include <uavcan/protocol/dynamic_node_id_server/centralized.hpp>
+#include <uavcan/protocol/HardwareVersion.hpp>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include <AP_BoardConfig/AP_BoardConfig_CAN.h>
+
+#include <AP_Common/AP_Common.h>
+#include <DataFlash/DataFlash.h>
+
+#if HAL_OS_POSIX_IO
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <dirent.h>
+#if defined(__APPLE__) && defined(__MACH__)
+#include <sys/param.h>
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#endif
+#endif
+
+#if HAL_OS_FATFS_IO
+#include <stdio.h>
+#endif
+
+#ifndef UAVCAN_NODE_DB_PATH
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+#define UAVCAN_NODE_DB_PATH "/fs/microsd/APM/UAVCAN"
+#else
+#define UAVCAN_NODE_DB_PATH "/APM/UAVCAN"
```

Comment:
I just remembered, we have to define this in Linux board file (AP_HAL/boards/linux.h) to be inside the ArduPilot state dir by default.

Reply:
ok, can you push the change to this PR to make that happen?

Reply:
I've pushed a change to the PR to add HAL_BOARD_STORAGE_DIR to all HALs, and switched UAVCAN to use that